File: 0039-combination-sum/0039-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const results = [];
    candidates.sort((a,b) => a-b);
    const solver = (choices, target, pos) => {
        if (target == 0) return results.push(choices);
        if (target < candidates[0]) return;

        for (let i=pos;i<candidates.length;i++) {
            const choice = candidates[i];
            if (target - choice >= 0) solver([...choices,choice], target - choice, i);
        }
    }

    solver([],target,0);

    return results;
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        results = []
        def solver(curr_sum, used, i):
            if curr_sum == target:
                results.append(used.copy())
                return
            elif curr_sum > target or i == len(candidates):
                return
            else:
                used.append(candidates[i])
                solver(curr_sum + candidates[i], used, i)
                used.pop()
                solver(curr_sum, used, i+1)

        solver(0,[],0)
        return results
            
